prow/crier/reporters/gcs/kubernetes/reporter.go

```diff
@@ -0,0 +1,181 @@
+	podClientSets map[string]corev1.CoreV1Interface
+}
+
+func (rg k8sResourceGetter) GetPod(cluster, namespace, name string) (*v1.Pod, error) {
+	return rg.podClientSets[cluster].Pods(namespace).Get(name, metav1.GetOptions{})
+}
+
+func (rg k8sResourceGetter) GetEvents(cluster, namespace string, pod *v1.Pod) ([]v1.Event, error) {
+	events, err := rg.podClientSets[cluster].Events(namespace).Search(scheme.Scheme, pod)
+	if err != nil {
+		return nil, err
+	}
+	return events.Items, nil
+}
+
+func (gr *gcsK8sReporter) Report(pj *prowv1.ProwJob) ([]*prowv1.ProwJob, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second) // TODO: pass through a global context?
+	defer cancel()
+
+	_, _, err := util.GetJobDestination(gr.cfg, pj)
+	if err != nil {
+		gr.logger.Infof("Not uploading %q (%s#%s) because we couldn't find a destination: %v", pj.Name, pj.Spec.Job, pj.Status.BuildID, err)
+		return []*prowv1.ProwJob{pj}, nil
+	}
+
+	return []*prowv1.ProwJob{pj}, gr.reportPodInfo(ctx, pj)
+}
+
+func (gr *gcsK8sReporter) reportPodInfo(ctx context.Context, pj *prowv1.ProwJob) error {
+	// We only report this after a prowjob is complete (and, therefore, pod state is immutable)
+	if !pj.Complete() {
+		return errors.New("cannot report incomplete jobs")
+	}
+
+	pod, err := gr.rg.GetPod(pj.Spec.Cluster, gr.cfg().PodNamespace, pj.Name)
+	if err != nil {
+		// If we return an error we will be retried ~indefinitely. Given that permanent errors
+		// are expected (pods will be garbage collected), this isn't useful. Instead, just
+		// go along with it.
+		gr.logger.WithError(err).Infof("Couldn't fetch info for pod %s", pj.Name)
+		pod = nil
+	}
+
+	var events []v1.Event
+	if pod != nil {
+		events, err = gr.rg.GetEvents(pj.Spec.Cluster, gr.cfg().PodNamespace, pod)
+		if err != nil {
+			gr.logger.WithError(err).Infof("Couldn't fetch events for pod %s", pj.Name)
+		}
+	}
+
+	if pod == nil && len(events) == 0 {
+		gr.logger.Infof("Not reporting on job %q because we could fetch neither pod nor events", pj.Name)
+		return nil
+	}
+
+	report := PodReport{
+		Pod:    pod,
+		Events: events,
+	}
+
+	output, err := json.Marshal(report)
+	if err != nil {
+		// This should never happen.
+		gr.logger.WithError(err).Warn("Couldn't marshal pod info")
+	}
+
+	bucketName, dir, err := util.GetJobDestination(gr.cfg, pj)
+	if err != nil {
+		return fmt.Errorf("couldn't get job destination: %v", err)
+	}
+
+	if gr.dryRun {
+		gr.logger.Infof("Would upload pod info to %q/%q", bucketName, dir)
+		return nil
+	}
+
+	return util.WriteContent(ctx, gr.logger, gr.author, bucketName, path.Join(dir, "podinfo.json"), true, output)
+}
+
+func (gr *gcsK8sReporter) GetName() string {
+	return reporterName
+}
+
+func (gr *gcsK8sReporter) ShouldReport(pj *prowv1.ProwJob) bool {
+	// This reporting only makes sense for the Kubernetes agent (otherwise we don't
+	// have a pod to look up). It is only particularly useful for us to look at
+	// complete jobs.
```

Comment:
As discussed offline:

We can only do work on prowjob state transitions. In practice, we're unlikely to see stuck pods at any state transition except to a terminal one -- when "pending" the pod is probably not yet obviously stuck.

Pods that are stuck will eventually fail, at which point the pod will transition to the failed state, and we can pick up the pod information. It would probably be nice to reduce the pod pending timeout for k8s (it almost certainly does not need to be an hour), which would improve this experience. In the long term, we might want to embed the concept of prowjob health in its status.